pkg/controller/utils/datadog/datadog.go

```diff
@@ -0,0 +1,451 @@
+}
+
+// Start establishes a connection with the Datadog API
+// it starts sending deployment metrics once the connection is validated
+func (dd *MetricsForwarder) Start(stop <-chan struct{}) error {
+	// wait.PollInfinite is blocking until dd.connectToDatadogAPI returns true
+	if err := wait.PollInfinite(dd.retryInterval, dd.connectToDatadogAPI); err == nil {
+		log.Info("Datadog metrics forwarder initilized successfully")
+	}
+
+	if err := dd.sendStartupEvent(); err != nil {
+		log.Error(err, "cannot send operator startup event")
+	}
+
+	metricsTicker := time.NewTicker(dd.sendMetricsInterval)
+	defer metricsTicker.Stop()
+	for {
+		select {
+		case <-stop:
+			return nil
+		case <-metricsTicker.C:
+			if err := dd.forwardMetrics(); err != nil {
+				log.Error(err, "cannot forward metrics to Datadog")
+			}
+		}
+	}
+}
+
+// connectToDatadogAPI ensures the connection to the Datadog API is valid
+// implements wait.ConditionFunc and never returns error to keep retrying
+func (dd *MetricsForwarder) connectToDatadogAPI() (bool, error) {
+	dad, err := dd.listDatadogAgentDeployment()
+	if err != nil {
+		log.Error(err, "cannot list DatadogAgentDeploymentList to get Datadog credentials")
+		return false, nil
+	}
+	log.Info("Getting Datadog credentials")
+	apiKey, appKey, err := dd.getCredentials(dad)
+	if err != nil {
+		log.Error(err, "cannot get Datadog credentials")
+		return false, nil
+	}
+	log.Info("Initializing Datadog metrics forwarder")
+	if err := dd.initAPIClient(apiKey, appKey); err != nil {
+		log.Error(err, "cannot get Datadog metrics forwarder to send deployment metrics, will retry later...")
+		return false, nil
+	}
+	return true, nil
+}
+
+// forwardMetrics sends metrics to Datadog
+// it tries to refresh credentials each time it's called
+func (dd *MetricsForwarder) forwardMetrics() error {
+	dad, err := dd.listDatadogAgentDeployment()
+	if err != nil {
+		log.Error(err, "cannot list DatadogAgentDeploymentList to get deployment metrics")
+		return err
+	}
+	apiKey, appKey, err := dd.getCredentials(dad)
+	if err != nil {
+		log.Error(err, "cannot get Datadog credentials")
+		return err
+	}
+	if err := dd.updateCredsIfNeeded(apiKey, appKey); err != nil {
+		log.Error(err, "cannot update Datadog credentials")
+		return err
+	}
+	log.Info("Collecting metrics and events")
+	dd.setTags(dad)
+	status := dad.Status.DeepCopy()
+	if err := dd.sendStatusMetrics(status); err != nil {
+		log.Error(err, "cannot send status metrics to Datadog")
+		return err
+	}
+	if err := dd.sendStatusEvents(status); err != nil {
+		log.Error(err, "cannot send events to Datadog")
+		return err
+	}
+	return nil
+}
+
+// sendStartupEvent posts the operator startup event
+// it must be called only once, when the operator starts
+func (dd *MetricsForwarder) sendStartupEvent() error {
+	event := &api.Event{
+		Time:      api.Int(int(time.Now().Unix())),
+		Title:     api.String("Datadog Operator has started"),
+		EventType: api.String("Datadog Operator Startup"),
+	}
```

Comment:
we should add `SourceType ` too for a better integration in the datadog website

Reply:
good catch thanks